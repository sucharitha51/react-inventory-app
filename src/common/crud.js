import initialData from '../initialData';

export const setData = (itemData = initialData) => {
  localStorage.setItem('itemData', JSON.stringify(itemData));
};

export const loadData = () => {
  const localStorageExists = localStorage.hasOwnProperty('itemData');
  if (!localStorageExists) setData();
  return JSON.parse(localStorage.getItem('itemData'));
};

export const deleteItem = (id, itemsList) => {
  const updatedList = itemsList.filter(item => item.id !== id);

  setData(updatedList);
};

export const addItem = (newItem, itemsList) => {
  const updatedList = [
    ...itemsList,
    newItem
  ];

  setData(updatedList);
};

export const updateItem = (updatedItem, itemsList) => {
  const updatedList = [...itemsList];
  const itemIndex = itemsList.findIndex(item => item.id === updatedItem.id);
  updatedList[itemIndex] = updatedItem;

  setData(updatedList);
};