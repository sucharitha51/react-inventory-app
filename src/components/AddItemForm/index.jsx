import React, { useState } from 'react';

import 'react-responsive-modal/styles.css';
import { Modal } from 'react-responsive-modal';

import '../../styles/_modal.scss';
import errorIcon from '../../assets/error.svg';

import { addItem } from '../../common/crud';
import { onlySpaces } from '../../common/regex';

const AddItemForm = ({ openAddNewModal, setOpenAddNewModal, itemsList, fetchAndSetData }) => {
  const [itemName, setItemName] = useState('');
  const [itemSerialNumber, setItemSerialNumber] = useState('');
  const [itemPrice, setItemPrice] = useState('');
  const [errorMsg, setErrorMsg] = useState('');

  const handleInputChange = e => {
    if (e.target.name === 'itemName') setItemName(e.target.value);
    if (e.target.name === 'itemSerialNumber') setItemSerialNumber(e.target.value);
    if (e.target.name === 'itemPrice') setItemPrice(e.target.value);
  };

  const handleAddItem = e => {
    e.preventDefault();

    if (onlySpaces.test(itemName) || onlySpaces.test(itemSerialNumber) || onlySpaces.test(itemPrice)) {
      setErrorMsg('Input fields can\'t be empty.');
    } else {
      const lastItemIdInList = Math.max.apply(Math, itemsList.map(item => item.id));
      const newId = lastItemIdInList === -Infinity ? 1 : lastItemIdInList + 1;

      const newItem = {
        id: newId,
        name: itemName.trim(),
        serialNumber: itemSerialNumber,
        price: itemPrice
      };

      addItem(newItem, itemsList);
      fetchAndSetData();
      resetForm();
    }
  };

  const resetForm = () => {
    setItemName('');
    setItemSerialNumber('');
    setItemPrice('');
    setErrorMsg('');
    setOpenAddNewModal(false);
  };

  return (
    <Modal classNames={{ modal: 'modal' }} open={openAddNewModal} onClose={() => setOpenAddNewModal(false)} center>
      <h2 className='title'>Add New Item</h2>
      {errorMsg && <p className='error-message'><img src={errorIcon} alt='Error' /> {errorMsg}</p>}

      <form onSubmit={handleAddItem}>
        <label htmlFor='newArtist'>Name:</label>
        <input
          type='text'
          name='itemName'
          placeholder='Item Name'
          value={itemName}
          onChange={handleInputChange}
          required={true}
        />

        <label htmlFor='serialNumber'>Serial Number:</label>
        <input
          type='text'
          name='itemSerialNumber'
          placeholder='Serial Number'
          value={itemSerialNumber}
          onChange={handleInputChange}
          required={true}
        />

        <label htmlFor='price'>Value:</label>
        <input
          type='number'
          name='itemPrice'
          placeholder='Price'
          value={itemPrice}
          onChange={handleInputChange}
          required={true}
        />

        <button className='confirm' type='submit'>Confirm</button>
      </form>
      <button className='cancel' type='reset' onClick={() => setOpenAddNewModal(false)}>Cancel</button>
    </Modal>
  )
}

export default AddItemForm;