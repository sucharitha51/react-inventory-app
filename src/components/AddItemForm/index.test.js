import { configure, shallow } from 'enzyme'
import React from 'react'

import Adapter from 'enzyme-adapter-react-16'
import AddItemForm from './index';
import * as crud from '../../common/crud';
configure({ adapter: new Adapter() })

let [setOpenAddNewModal, fetchAndSetData, resetForm] = new Array(2).fill(jest.fn());

let wrapper
const props = { openAddNewModal: true, setOpenAddNewModal, itemsList: [], fetchAndSetData }

beforeEach(() => {
    wrapper = shallow(<AddItemForm {...props} />)
})

describe('Add item form', () => {

    it('should render a add item form', () => {
        expect(wrapper.find('input').at(0).props().defaultValue).toEqual(props.name);
        expect(wrapper.find('input').at(1).props().defaultValue).toEqual(props.serialNumber);
        expect(wrapper.find('input').at(2).props().defaultValue).toEqual(props.price);
    });

    let container = shallow(<AddItemForm />);
    it('should match the snapshot', () => {
        expect(container.html()).toMatchSnapshot();
    });

    it('should have a name field', () => {
        expect(wrapper.find('input[name="itemName"]').length).toEqual(1);
    });

    it('should have proper props for name field', () => {
        expect(wrapper.find('input[name="itemName"]').props()).toEqual({
            type: 'text',
            name: 'itemName',
            placeholder: 'Item Name',
            value: '',
            onChange: expect.anything(),
            required: true
        });
    });

    it('should have a serial number field', () => {
        expect(wrapper.find('input[name="itemSerialNumber"]').length).toEqual(1);
    });

    it('should have proper props for Serial Number field', () => {
        expect(wrapper.find('input[name="itemSerialNumber"]').props()).toEqual({
            type: 'text',
            name: 'itemSerialNumber',
            placeholder: 'Serial Number',
            value: '',
            onChange: expect.anything(),
            required: true
        });
    });

    it('should have a price field', () => {
        expect(wrapper.find('input[name="itemSerialNumber"]').length).toEqual(1);
    });

    it('should have proper props for Price field', () => {
        expect(wrapper.find('input[name="itemPrice"]').props()).toEqual({
            type: 'number',
            name: 'itemPrice',
            placeholder: 'Price',
            value: '',
            onChange: expect.anything(),
            required: true
        });
    });

    it('should name field have value', () => {
        const name = wrapper.find('input[name="itemName"]');
        name.simulate('change', { target: { name: 'itemName', value: "Test" } });
        expect(wrapper.find('input[name="itemName"]').prop('value')).toEqual("Test")
    });

    it('should serial number field have value', () => {
        const serialNumber = wrapper.find('input[name="itemSerialNumber"]');
        serialNumber.simulate('change', { target: { name: 'itemSerialNumber', value: "12345Test" } });
        expect(wrapper.find('input[name="itemSerialNumber"]').prop('value')).toEqual("12345Test")
    });

    it('should price field have value', () => {
        const itemPrice = wrapper.find('input[name="itemPrice"]');
        itemPrice.simulate('change', { target: { name: 'itemPrice', value: 100 } });
        expect(wrapper.find('input[name="itemPrice"]').prop('value')).toEqual(100)
    });

    it('should have a submit button', () => {
        expect(wrapper.find('button[type="submit"]').length).toEqual(1);
    });

    it('should have a cancel button', () => {
        expect(wrapper.find('button[type="reset"]').length).toEqual(1);
    });

    it('it should add a new item', () => {
        const fakeEvent = { preventDefault: () => console.log('preventDefault') };

        const name = wrapper.find('input[name="itemName"]');
        name.simulate('change', { target: { name: 'itemName', value: "Test" } });
        expect(wrapper.find('input[name="itemName"]').prop('value')).toEqual("Test")
  
        const serialNumber = wrapper.find('input[name="itemSerialNumber"]');
        serialNumber.simulate('change', { target: { name: 'itemSerialNumber', value: "12345Test" } });
        expect(wrapper.find('input[name="itemSerialNumber"]').prop('value')).toEqual("12345Test")

        const itemPrice = wrapper.find('input[name="itemPrice"]');
        itemPrice.simulate('change', { target: { name: 'itemPrice', value: 100 } });
        expect(wrapper.find('input[name="itemPrice"]').prop('value')).toEqual(100)

        expect(wrapper.find('form').length).toBe(1);
    
        jest.spyOn(crud, 'addItem').mockImplementation(() => Promise.resolve({message: "Item has been added"}))
  
        wrapper.find('form').simulate('submit', fakeEvent)

        expect(wrapper.find('input[name="itemName"]').prop('value')).toEqual("")  
      
        expect(crud.addItem).toHaveBeenCalledWith({id: 1, name: 'Test', price: 100, serialNumber: '12345Test'}, [])
        expect(fetchAndSetData).toHaveBeenCalled()
    });

});

