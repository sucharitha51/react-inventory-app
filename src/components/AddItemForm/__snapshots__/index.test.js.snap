// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add item form should match the snapshot 1`] = `""`;
